File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/verify/ResetPasswordServiceImpl.java
package uk.gov.justice.digital.hmpps.oauth2server.verify;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.LockedException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken.TokenType;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository;
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserTokenRepository;
import uk.gov.justice.digital.hmpps.oauth2server.nomis.model.NomisUserPersonDetails;
import uk.gov.justice.digital.hmpps.oauth2server.security.UserPersonDetails;
import uk.gov.justice.digital.hmpps.oauth2server.security.UserService;
import uk.gov.justice.digital.hmpps.oauth2server.service.DelegatingUserService;
import uk.gov.justice.digital.hmpps.oauth2server.utils.EmailHelper;
import uk.gov.service.notify.NotificationClientApi;
import uk.gov.service.notify.NotificationClientException;

import java.util.Collections;
import java.util.Map;
import java.util.Optional;

import static uk.gov.justice.digital.hmpps.oauth2server.security.AuthSource.nomis;
import static uk.gov.justice.digital.hmpps.oauth2server.security.AuthSource.valueOf;

@Service
@Slf4j
@Transactional
public class ResetPasswordServiceImpl implements ResetPasswordService, PasswordService {

    private final UserRepository userRepository;
    private final UserTokenRepository userTokenRepository;
    private final UserService userService;
    private final DelegatingUserService delegatingUserService;
    private final NotificationClientApi notificationClient;
    private final String resetTemplateId;
    private final String resetUnavailableTemplateId;
    private final String resetUnavailableEmailNotFoundTemplateId;
    private final String resetPasswordConfirmedTemplateId;

    public ResetPasswordServiceImpl(final UserRepository userRepository,
                                    final UserTokenRepository userTokenRepository,
                                    final UserService userService,
                                    final DelegatingUserService delegatingUserService,
                                    final NotificationClientApi notificationClient,
                                    @Value("${application.notify.reset.template}") final String resetTemplateId,
                                    @Value("${application.notify.reset-unavailable.template}") final String resetUnavailableTemplateId,
                                    @Value("${application.notify.reset-unavailable-email-not-found.template}") final String resetUnavailableEmailNotFoundTemplateId,
                                    @Value("${application.notify.reset-password.template}") final String resetPasswordConfirmedTemplateId) {
        this.userRepository = userRepository;
        this.userTokenRepository = userTokenRepository;
        this.userService = userService;
        this.delegatingUserService = delegatingUserService;
        this.notificationClient = notificationClient;
        this.resetTemplateId = resetTemplateId;
        this.resetUnavailableTemplateId = resetUnavailableTemplateId;
        this.resetUnavailableEmailNotFoundTemplateId = resetUnavailableEmailNotFoundTemplateId;
        this.resetPasswordConfirmedTemplateId = resetPasswordConfirmedTemplateId;
    }

    @Override
    @Transactional(transactionManager = "authTransactionManager")
    public Optional<String> requestResetPassword(final String usernameOrEmailAddress, final String url) throws NotificationClientRuntimeException {
        final Optional<User> optionalUser;
        final boolean multipleMatchesAndCanBeReset;
        if (StringUtils.contains(usernameOrEmailAddress, "@")) {
            final var email = EmailHelper.format(usernameOrEmailAddress);
            final var matches = userRepository.findByEmail(email);
            if (matches.isEmpty()) {
                // no match, but got an email address so let them know
                sendEmail(email, resetUnavailableEmailNotFoundTemplateId, Collections.emptyMap(), email);
                return Optional.empty();
            }

            final var passwordCanBeReset = matches.stream().filter(this::passwordAllowedToBeReset).findFirst();
            multipleMatchesAndCanBeReset = passwordCanBeReset.map(ue -> matches.size() > 1).orElse(Boolean.FALSE);
            optionalUser = Optional.of(matches.get(0));
        } else {
            multipleMatchesAndCanBeReset = false;

            optionalUser = userRepository.findByUsername(usernameOrEmailAddress.toUpperCase())
                    .or(() -> userService.findMasterUserPersonDetails(usernameOrEmailAddress.toUpperCase()).flatMap(userPersonDetails -> {

                        switch (valueOf(userPersonDetails.getAuthSource())) {
                            case nomis:
                                return saveNomisUser(userPersonDetails);
                            case delius:
                                return saveDeliusUser(userPersonDetails);
                        }
                        return Optional.empty();
                    }));
        }

        if (optionalUser.isEmpty() || StringUtils.isEmpty(optionalUser.get().getEmail())) {
            // no user found or email address found, so nothing more we can do.  Bail
            return Optional.empty();
        }

        final var user = optionalUser.get();
        final var templateAndParameters = getTemplateAndParameters(url, multipleMatchesAndCanBeReset, user);

        sendEmail(user.getUsername(), templateAndParameters, user.getEmail());

        return Optional.ofNullable(templateAndParameters.getResetLink());
    }

    private Optional<User> saveDeliusUser(final UserPersonDetails userPersonDetails) {
        final var user = userPersonDetails.toUser();
        if (!passwordAllowedToBeReset(user, userPersonDetails)) {
            return Optional.empty();
        }
        userRepository.save(user);
        return Optional.of(user);
    }

    private Optional<User> saveNomisUser(final UserPersonDetails userPersonDetails) {
        final var user = userPersonDetails.toUser();
        if (!passwordAllowedToBeReset(user, userPersonDetails)) {
            return Optional.empty();
        }
        return userService.getEmailAddressFromNomis(user.getUsername()).map(e -> {
            user.setEmail(e);
            userRepository.save(user);
            return user;
        });
    }

    private TemplateAndParameters getTemplateAndParameters(final String url, final boolean multipleMatchesAndCanBeReset, final User user) {
        final UserPersonDetails userDetails;
        if (user.isMaster()) {
            userDetails = user;
        } else {
            final var userOptional = userService.findMasterUserPersonDetails(user.getUsername());
            if (userOptional.isEmpty()) {
                // shouldn't really happen, means that a nomis user exists in auth but not in nomis
                return new TemplateAndParameters(resetUnavailableTemplateId, user.getUsername(), user.getName());
            }
            userDetails = userOptional.get();
        }
        // only allow reset for active accounts that aren't locked
        // or are locked by getting password incorrect (in either c-nomis or auth)
        final var firstName = userDetails.getFirstName();
        final var fullName = userDetails.getName();
        if (multipleMatchesAndCanBeReset || passwordAllowedToBeReset(user, userDetails)) {
            final var userToken = user.createToken(TokenType.RESET);

            final var selectOrConfirm = multipleMatchesAndCanBeReset ? "select" : "confirm";
            final var resetLink = String.format("%s-%s?token=%s", url, selectOrConfirm, userToken.getToken());
            return new TemplateAndParameters(resetTemplateId, Map.of("firstName", firstName, "fullName", fullName, "resetLink", resetLink));
        }
        return new TemplateAndParameters(resetUnavailableTemplateId, firstName, fullName);
    }

    private void sendEmail(final String username, final TemplateAndParameters templateAndParameters, final String email) throws NotificationClientRuntimeException {
        sendEmail(username, templateAndParameters.getTemplate(), templateAndParameters.getParameters(), email);
    }

    private void sendEmail(final String username, final String template, final Map<String, String> parameters, final String email) throws NotificationClientRuntimeException {
        try {
            log.info("Sending reset password to notify for user {}", username);
            notificationClient.sendEmail(template, email, parameters, null);
        } catch (final NotificationClientException e) {
            log.warn("Failed to send reset password to notify for user {}", username, e);
            if (e.getHttpResult() >= 500) {
                // second time lucky
                try {
                    notificationClient.sendEmail(template, email, parameters, null);
                } catch (final NotificationClientException e1) {
                    throw new NotificationClientRuntimeException(e1);
                }
            }
            throw new NotificationClientRuntimeException(e);
        }
    }

    private boolean passwordAllowedToBeReset(final User user, final UserPersonDetails userPersonDetails) {
        if (user.getSource() != nomis) {
            // for non nomis users they must be enabled (so can be locked)
            return userPersonDetails.isEnabled();
        }
        // otherwise must be nomis user so will have a staff account instead.
        final var staffUserAccount = (NomisUserPersonDetails) userPersonDetails;
        final var status = staffUserAccount.getAccountDetail().getStatus();
        return staffUserAccount.getStaff().isActive() && (!status.isLocked() || status.isUserLocked() || user.isLocked());
    }

    @Override
    @Transactional(transactionManager = "authTransactionManager")
    public void setPassword(final String token, final String password) {
        final var userToken = userTokenRepository.findById(token).orElseThrow();
        final var user = userToken.getUser();

        final var userPersonDetails = user.isMaster() ? user : userService.findMasterUserPersonDetails(user.getUsername()).orElseThrow();
        if (!passwordAllowedToBeReset(user, userPersonDetails)) {
            // failed, so let user know
            throw new LockedException("locked");
        }

        delegatingUserService.changePasswordWithUnlock(userPersonDetails, password);

        user.removeToken(userToken);
        userRepository.save(user);
        sendPasswordResetEmail(user);
    }

    private void sendPasswordResetEmail(final User user) {
        // then the reset token
        final var username = user.getUsername();
        final var email = user.getEmail();
        final var parameters = Map.of("firstName", user.getFirstName(), "fullName", user.getName(), "username", username);

        // send the email
        try {
            log.info("Sending password reset to notify for user {}", username);
            notificationClient.sendEmail(resetPasswordConfirmedTemplateId, email, parameters, null);
        } catch (final NotificationClientException e) {
            final var reason = (e.getCause() != null ? e.getCause() : e).getClass().getSimpleName();
            log.warn("Failed to send password reset notify for user {} due to {}", username, reason, e);
            if (e.getHttpResult() >= 500) {
                // second time lucky
                try {
                    notificationClient.sendEmail(resetPasswordConfirmedTemplateId, email, parameters, null, null);
                } catch (final NotificationClientException ex) {
                    log.error("Failed to send password reset notify for user {}", username, ex);
                }
            }
        }
    }

    private boolean passwordAllowedToBeReset(final User ue) {
        final var userPersonDetailsOptional = ue.isMaster() ? Optional.of(ue) : userService.findMasterUserPersonDetails(ue.getUsername());
        return userPersonDetailsOptional.map(upd -> passwordAllowedToBeReset(ue, upd)).orElse(false);
    }

    @Override
    @Transactional(transactionManager = "authTransactionManager")
    public String moveTokenToAccount(final String token, final String usernameInput) throws ResetPasswordException {
        if (StringUtils.isBlank(usernameInput)) {
            throw new ResetPasswordException("missing");
        }
        final var username = StringUtils.upperCase(StringUtils.trim(usernameInput));

        final var userOptional = userRepository.findByUsername(username);
        return userOptional.map(ue -> {
            final var userToken = userTokenRepository.findById(token).orElseThrow();
            // need to have same email address
            if (!userToken.getUser().getEmail().equals(ue.getEmail())) {
                throw new ResetPasswordException("email");
            }

            if (!passwordAllowedToBeReset(ue)) {
                throw new ResetPasswordException("locked");
            }
            if (userToken.getUser().getUsername().equals(username)) {
                // no work since they are the same
                return token;
            }
            // otherwise need to delete and add
            userTokenRepository.delete(userToken);
            final var newUserToken = ue.createToken(TokenType.RESET);
            return newUserToken.getToken();

        }).orElseThrow(() -> new ResetPasswordException("notfound"));
    }

    @AllArgsConstructor
    @Getter
    private static class TemplateAndParameters {
        private final String template;
        private final Map<String, String> parameters;

        TemplateAndParameters(final String template, final String firstName, final String fullName) {
            this.template = template;
            this.parameters = Map.of("firstName", firstName, "fullName", fullName);
        }


        private String getResetLink() {
            return parameters.get("resetLink");
        }
    }

    // Throttling doesn't allow checked exceptions to be thrown, hence wrapped in a runtime exception
    public static class NotificationClientRuntimeException extends RuntimeException {

        public NotificationClientRuntimeException(final NotificationClientException e) {
            super(e);
        }
    }

    public static class ResetPasswordException extends RuntimeException {
        private final String reason;

        public ResetPasswordException(final String reason) {
            super(String.format("Reset Password failed with reason: %s", reason));
            this.reason = reason;
        }

        public String getReason() {
            return this.reason;
        }
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/verify/ResetPasswordServiceImpl.kt
package uk.gov.justice.digital.hmpps.oauth2server.verify

import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.LockedException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserRepository
import uk.gov.justice.digital.hmpps.oauth2server.auth.repository.UserTokenRepository
import uk.gov.justice.digital.hmpps.oauth2server.nomis.model.NomisUserPersonDetails
import uk.gov.justice.digital.hmpps.oauth2server.security.AuthSource
import uk.gov.justice.digital.hmpps.oauth2server.security.UserPersonDetails
import uk.gov.justice.digital.hmpps.oauth2server.security.UserService
import uk.gov.justice.digital.hmpps.oauth2server.service.DelegatingUserService
import uk.gov.justice.digital.hmpps.oauth2server.utils.EmailHelper.format
import uk.gov.service.notify.NotificationClientApi
import uk.gov.service.notify.NotificationClientException
import java.util.Optional

@Service
@Transactional
class ResetPasswordServiceImpl(
  private val userRepository: UserRepository,
  private val userTokenRepository: UserTokenRepository,
  private val userService: UserService,
  private val delegatingUserService: DelegatingUserService,
  private val notificationClient: NotificationClientApi,
  @Value("\${application.notify.reset.template}") private val resetTemplateId: String,
  @Value("\${application.notify.reset-unavailable.template}") private val resetUnavailableTemplateId: String,
  @Value("\${application.notify.reset-unavailable-email-not-found.template}") private val resetUnavailableEmailNotFoundTemplateId: String,
  @Value("\${application.notify.reset-password.template}") private val resetPasswordConfirmedTemplateId: String,
) : ResetPasswordService, PasswordService {

  companion object {
    private val log = LoggerFactory.getLogger(this::class.java)
  }

  @Transactional(transactionManager = "authTransactionManager")
  @Throws(NotificationClientRuntimeException::class)
  override fun requestResetPassword(usernameOrEmailAddress: String, url: String): Optional<String> {
    val optionalUser: Optional<User>
    val multipleMatchesAndCanBeReset: Boolean
    if (StringUtils.contains(usernameOrEmailAddress, "@")) {
      val email = format(usernameOrEmailAddress)
      val matches = userRepository.findByEmail(email)
      if (matches.isEmpty()) {
        // no match, but got an email address so let them know
        sendEmail(email, resetUnavailableEmailNotFoundTemplateId, emptyMap<String, String>(), email)
        return Optional.empty()
      }

      val passwordCanBeReset = matches.firstOrNull { passwordAllowedToBeReset(it) }
      multipleMatchesAndCanBeReset = passwordCanBeReset?.let { matches.size > 1 } ?: false
      optionalUser = Optional.of(matches[0])
    } else {
      multipleMatchesAndCanBeReset = false
      optionalUser = userRepository.findByUsername(usernameOrEmailAddress.toUpperCase())
        .or {
          userService.findMasterUserPersonDetails(usernameOrEmailAddress.toUpperCase())
            .flatMap { userPersonDetails: UserPersonDetails ->
              when (AuthSource.fromNullableString(userPersonDetails.authSource)) {
                AuthSource.nomis -> return@flatMap saveNomisUser(userPersonDetails)
                AuthSource.delius -> return@flatMap saveDeliusUser(userPersonDetails)
                else -> Optional.empty()
              }
            }
        }
    }
    if (optionalUser.isEmpty || StringUtils.isEmpty(optionalUser.get().email)) {
      // no user found or email address found, so nothing more we can do.  Bail
      return Optional.empty()
    }
    val user = optionalUser.get()
    val templateAndParameters = getTemplateAndParameters(url, multipleMatchesAndCanBeReset, user)
    sendEmail(user.username, templateAndParameters, user.email!!)
    return Optional.ofNullable(templateAndParameters.resetLink)
  }

  private fun saveDeliusUser(userPersonDetails: UserPersonDetails): Optional<User> {
    val user = userPersonDetails.toUser()
    if (!passwordAllowedToBeReset(user, userPersonDetails)) {
      return Optional.empty()
    }
    userRepository.save(user)
    return Optional.of(user)
  }

  private fun saveNomisUser(userPersonDetails: UserPersonDetails): Optional<User> {
    val user = userPersonDetails.toUser()
    return if (!passwordAllowedToBeReset(user, userPersonDetails)) {
      Optional.empty()
    } else userService.getEmailAddressFromNomis(user.username).map { e: String? ->
      user.email = e
      user.verified = true
      userRepository.save(user)
      user
    }
  }

  private fun getTemplateAndParameters(
    url: String,
    multipleMatchesAndCanBeReset: Boolean,
    user: User,
  ): TemplateAndParameters {
    val userDetails: UserPersonDetails
    userDetails = if (user.isMaster) {
      user
    } else {
      val userOptional = userService.findMasterUserPersonDetails(user.username)
      if (userOptional.isEmpty) {
        // shouldn't really happen, means that a nomis user exists in auth but not in nomis
        return TemplateAndParameters(resetUnavailableTemplateId, user.username, user.name)
      }
      userOptional.get()
    }
    // only allow reset for active accounts that aren't locked
    // or are locked by getting password incorrect (in either c-nomis or auth)
    val firstName = userDetails.firstName
    val fullName = userDetails.name
    if (multipleMatchesAndCanBeReset || passwordAllowedToBeReset(user, userDetails)) {
      val userToken = user.createToken(UserToken.TokenType.RESET)
      val selectOrConfirm = if (multipleMatchesAndCanBeReset) "select" else "confirm"
      val resetLink = "$url-$selectOrConfirm?token=${userToken.token}"
      return TemplateAndParameters(
        resetTemplateId,
        mapOf("firstName" to firstName, "fullName" to fullName, "resetLink" to resetLink)
      )
    }
    return TemplateAndParameters(resetUnavailableTemplateId, firstName, fullName)
  }

  @Throws(NotificationClientRuntimeException::class)
  private fun sendEmail(username: String, templateAndParameters: TemplateAndParameters, email: String) {
    sendEmail(username, templateAndParameters.template, templateAndParameters.parameters, email)
  }

  @Throws(NotificationClientRuntimeException::class)
  private fun sendEmail(username: String?, template: String, parameters: Map<String, String?>, email: String?) {
    try {
      log.info("Sending reset password to notify for user {}", username)
      notificationClient.sendEmail(template, email, parameters, null)
    } catch (e: NotificationClientException) {
      log.warn("Failed to send reset password to notify for user {}", username, e)
      if (e.httpResult >= 500) {
        // second time lucky
        try {
          notificationClient.sendEmail(template, email, parameters, null)
        } catch (e1: NotificationClientException) {
          throw NotificationClientRuntimeException(e1)
        }
      }
      throw NotificationClientRuntimeException(e)
    }
  }

  private fun passwordAllowedToBeReset(user: User, userPersonDetails: UserPersonDetails): Boolean {
    if (user.source != AuthSource.nomis) {
      // for non nomis users they must be enabled (so can be locked)
      return userPersonDetails.isEnabled
    }
    // otherwise must be nomis user so will have a staff account instead.
    val staffUserAccount = userPersonDetails as NomisUserPersonDetails
    val status = staffUserAccount.accountDetail.status
    return staffUserAccount.staff.isActive && (!status.isLocked || status.isUserLocked || user.locked)
  }

  @Transactional(transactionManager = "authTransactionManager")
  override fun setPassword(token: String, password: String?) {
    val userToken = userTokenRepository.findById(token).orElseThrow()
    val user = userToken.user
    val userPersonDetails =
      if (user.isMaster) user else userService.findMasterUserPersonDetails(user.username).orElseThrow()
    if (!passwordAllowedToBeReset(user, userPersonDetails)) {
      // failed, so let user know
      throw LockedException("locked")
    }
    delegatingUserService.changePasswordWithUnlock(userPersonDetails, password!!)
    user.removeToken(userToken)
    userRepository.save(user)
    sendPasswordResetEmail(user)
  }

  private fun sendPasswordResetEmail(user: User) {
    // then the reset token
    val username = user.username
    val email = user.email
    val parameters = mapOf("firstName" to user.firstName, "fullName" to user.name, "username" to username)

    // send the email
    try {
      log.info("Sending password reset to notify for user {}", username)
      notificationClient.sendEmail(resetPasswordConfirmedTemplateId, email, parameters, null)
    } catch (e: NotificationClientException) {
      val reason = (if (e.cause != null) e.cause else e)!!.javaClass.simpleName
      log.warn("Failed to send password reset notify for user {} due to {}", username, reason, e)
      if (e.httpResult >= 500) {
        // second time lucky
        try {
          notificationClient.sendEmail(resetPasswordConfirmedTemplateId, email, parameters, null, null)
        } catch (ex: NotificationClientException) {
          log.error("Failed to send password reset notify for user {}", username, ex)
        }
      }
    }
  }

  private fun passwordAllowedToBeReset(ue: User): Boolean {
    val userPersonDetailsOptional =
      if (ue.isMaster) Optional.of(ue) else userService.findMasterUserPersonDetails(ue.username)
    return userPersonDetailsOptional.map { upd: UserPersonDetails -> passwordAllowedToBeReset(ue, upd) }.orElse(false)
  }

  @Transactional(transactionManager = "authTransactionManager")
  @Throws(
    ResetPasswordException::class
  )
  override fun moveTokenToAccount(token: String, usernameInput: String?): String {
    if (StringUtils.isBlank(usernameInput)) {
      throw ResetPasswordException("missing")
    }
    val username = StringUtils.upperCase(StringUtils.trim(usernameInput))
    val userOptional = userRepository.findByUsername(username)
    return userOptional.map { ue: User ->
      val userToken = userTokenRepository.findById(token).orElseThrow()
      // need to have same email address
      if (userToken.user.email != ue.email) {
        throw ResetPasswordException("email")
      }
      if (!passwordAllowedToBeReset(ue)) {
        throw ResetPasswordException("locked")
      }
      if (userToken.user.username == username) {
        // no work since they are the same
        return@map token
      }
      // otherwise need to delete and add
      userTokenRepository.delete(userToken)
      val newUserToken = ue.createToken(UserToken.TokenType.RESET)
      newUserToken.token
    }.orElseThrow { ResetPasswordException("notfound") }
  }

  internal data class TemplateAndParameters(
    internal val template: String,
    internal val parameters: Map<String, String>,
  ) {
    constructor(template: String, firstName: String, fullName: String) :
      this(template, mapOf("firstName" to firstName, "fullName" to fullName))

    internal val resetLink: String?
      get() = parameters["resetLink"]
  }

  // Throttling doesn't allow checked exceptions to be thrown, hence wrapped in a runtime exception
  class NotificationClientRuntimeException(e: NotificationClientException?) : RuntimeException(e)
  class ResetPasswordException(val reason: String) : RuntimeException("Reset Password failed with reason: $reason")
}
